namui-cli/src/build/build_server/code_watcher.rs

```diff
@@ -0,0 +1,167 @@
+    watcher: INotifyWatcher,
+    context: Arc<Mutex<CodeWatcherContext>>,
+    watching_paths: HashSet<String>,
+    manifest_path: String,
+}
+
+impl CodeWatcher {
+    pub fn new(manifest_path: String) -> Self {
+        let context = Arc::new(Mutex::new(CodeWatcherContext {
+            was_changed: true,
+            sender: None,
+        }));
+        let mut root_watching_path = PathBuf::from(&manifest_path);
+        root_watching_path.pop();
+        let root_watching_path = root_watching_path.join("src").to_string_lossy().to_string();
+
+        let (watcher_sender, watcher_receiver) = std::sync::mpsc::channel::<DebouncedEvent>();
+        let mut watcher = notify::watcher(watcher_sender, Duration::from_secs(1)).unwrap();
+        if let Err(error) = watcher.watch(root_watching_path, RecursiveMode::Recursive) {
+            panic!("watch failed {:?}", error);
+        }
+
+        let context_clone = context.clone();
+        thread::spawn(move || -> ! {
+            loop {
+                match watcher_receiver.recv() {
+                    Ok(event) => {
+                        match event {
+                            DebouncedEvent::Create(_)
+                            | DebouncedEvent::Remove(_)
+                            | DebouncedEvent::Rename(_, _)
+                            | DebouncedEvent::Write(_) => {
+                                debug_println!("CodeWatcher: locking code_watcher.context...");
+                                let mut context = block_on(context_clone.lock());
+                                debug_println!("CodeWatcher: code_watcher.context locked");
+
+                                match context.sender.take() {
+                                    Some(sender) => {
+                                        if let Err(_) = sender.send(()) {
+                                            panic!("send failed in watcher")
+                                        }
+                                    }
+                                    None => {
+                                        context.was_changed = true;
+                                    }
+                                }
+                            }
+                            _ => (),
+                        };
+                    }
+                    Err(error) => eprintln!("{:?}", error),
+                }
+            }
+        });
+
+        Self {
+            watcher,
+            context,
+            watching_paths: HashSet::new(),
+            manifest_path,
+        }
+    }
+
+    pub async fn wait_for_change(&self) {
+        match self.check_changed_from_last_check().await {
+            true => (),
+            false => {
+                let receiver = self.register_sender().await;
+                match receiver.await {
+                    Ok(_) => (),
+                    Err(error) => panic!("code_watcher oneshot receive error: {:?}", error),
+                }
+            }
+        }
```

Comment:
If there was changed after last check. It will return `()` immediately.
Else, It will wait for code change.

Returning `()` immediately means "There was code change while you rebuilding!"

Reply:
I think it could make a timing issue. What if other thread found changes between 79 line and 82 line?

Reply:
Oh!. thx!
Timing issue is likely to be fixed by this commit 23ca59f.